face-detection/src/mediapipe/detector.ts

```diff
@@ -0,0 +1,156 @@
+/**
+ * @license
+ * Copyright 2021 Google LLC. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * =============================================================================
+ */
+import * as faceDetection from '@mediapipe/face_detection';
+import * as tf from '@tensorflow/tfjs-core';
+
+import {MEDIAPIPE_FACE_DETECTOR_KEYPOINTS} from '../constants';
+import {FaceDetector} from '../face_detector';
+import {getBoundingBox} from '../shared/calculators/association_norm_rect';
+import {Keypoint} from '../shared/calculators/interfaces/common_interfaces';
+import {Face, FaceDetectorInput} from '../types';
+
+import {validateModelConfig} from './detector_utils';
+import {MediaPipeFaceDetectorMediaPipeEstimationConfig, MediaPipeFaceDetectorMediaPipeModelConfig} from './types';
+
+/**
+ * MediaPipe detector class.
+ */
+export class MediaPipeFaceDetectorMediaPipe implements FaceDetector {
+  private readonly faceDetectorSolution: faceDetection.FaceDetection;
+
+  // This will be filled out by asynchronous calls to onResults. They will be
+  // stable after `await send` is called on the faces solution.
+  private width = 0;
+  private height = 0;
+  private faces: Face[];
+
+  private selfieMode = false;
+
+  // Should not be called outside.
+  constructor(config: MediaPipeFaceDetectorMediaPipeModelConfig) {
+    this.faceDetectorSolution = new faceDetection.FaceDetection({
+      locateFile: (path, base) => {
+        if (config.solutionPath) {
+          const solutionPath = config.solutionPath.replace(/\/+$/, '');
+          return `${solutionPath}/${path}`;
+        }
+        return `${base}/${path}`;
+      }
+    });
+    this.faceDetectorSolution.setOptions(
+        {selfieMode: this.selfieMode, model: config.modelType});
+    this.faceDetectorSolution.onResults(results => {
+      this.height = results.image.height;
+      this.width = results.image.width;
+      this.faces = [];
+      if (results.detections !== null) {
+        for (const detection of results.detections.slice(0, config.maxFaces)) {
+          this.faces.push({
+            keypoints: this.translateOutput(detection.landmarks),
+            box: getBoundingBox(detection.boundingBox)
+          });
+        }
+      }
+    });
+  }
+
+  private translateOutput(landmarks: faceDetection.NormalizedLandmarkList) {
```

Comment:
It's not exactly "translate". You are converting normalized coordinates to absolute.